rllib/policy/eager_tf_policy.py

```diff
@@ -156,54 +148,70 @@ def compute_actions_from_input_dict(
-            """
-            # Have we traced the `_compute_gradients_eager` function yet?
-            if self._traced_compute_gradients is None:
-                self._traced_compute_gradients = convert_eager_inputs(
+            # Create a traced version of `self._compute_gradients_helper`.
+            if self._traced_compute_gradients_helper is False:
+                self._compute_gradients_helper = convert_eager_inputs(
                     tf.function(
-                        super()._compute_gradients_eager,
+                        super()._compute_gradients_helper,
                         autograph=False,
-                        experimental_relax_shapes=True))
+                        experimental_relax_shapes=False))
+                self._traced_compute_gradients_helper = True
 
-            # Call the only-once compiled traced function with the SampleBatch
-            # (will be converted to tensors beforehand).
-            return self._traced_compute_gradients(samples)
+            # Now that the helper method is traced, call super's
+            # apply_gradients (which will call the traced helper).
+            return super(TracedEagerPolicy, self).compute_gradients(samples)
 
         @override(Policy)
-        @convert_eager_outputs
         def apply_gradients(self, grads: ModelGradients) -> None:
-            """Traced version of EagerTFPolicy's `apply_gradients` method.
-
-            Args:
-                grads: The ModelGradients to apply using our optimizer(s).
-            """
-            # Have we traced the `apply_gradients` function yet?
-            if self._traced_apply_gradients is None:
-                self._traced_apply_gradients = tf.function(
-                    super().apply_gradients,
-                    autograph=False,
-                    experimental_relax_shapes=True)
-
-            # Call the only-once compiled traced function with the grads
-            # (will be converted to tensors beforehand).
-            return self._traced_apply_gradients(grads)
-
-    TracedEagerPolicy.__name__ = eager_policy_cls.__name__
-    TracedEagerPolicy.__qualname__ = eager_policy_cls.__qualname__
+            """Traced version of Policy.apply_gradients."""
+
+            self._check_too_many_retraces()
+
+            # Create a traced version of `self._apply_gradients_helper`.
+            if self._traced_apply_gradients_helper is False:
+                self._traced_apply_gradients = convert_eager_inputs(
+                    tf.function(
+                        super()._apply_gradients_helper,
+                        autograph=False,
+                        experimental_relax_shapes=False))
+                self._traced_apply_gradients_helper = True
+
+            # Now that the helper method is traced, call super's
+            # apply_gradients (which will call the traced helper).
+            return super(TracedEagerPolicy, self).apply_gradients(grads)
+
+        def _check_too_many_retraces(self):
```

Comment:
can you add a function description here about when we should add this function to enforce retracing counts? in case we are going to add additional apis that may cause problems.

one random idea I have is maybe we can make this a private decorator in this function, so instead of writing:

```
def compute_gradient(self):
    self._check_too_many_retraces()
```

we can write:

```
@check_too_many_retraces()
def compute_gradient(self):
    .....
```

no functional change. just read slightly better. maybe. not important though.

Reply:
+1

Reply:
done